sdk/src/main/scala/com/horizen/account/websocket/WebSocketAccountServerEndpoint.scala

```diff
@@ -0,0 +1,308 @@
+  private val mapper = new ObjectMapper().registerModule(DefaultScalaModule)
+
+  @OnClose
+  def onClose(session: Session): Unit = {
+    log.debug("Websocket session closed "+session.getId)
+    WebSocketAccountServerEndpoint.removeSession(session)
+  }
+
+  @OnError
+  def onError(session: Session, t:Throwable): Unit = {
+    log.error("Error on websocket session "+session.getId+": "+t.toString)
+    WebSocketAccountServerEndpoint.removeSession(session)
+  }
+
+  @OnMessage
+  def onMessageReceived(session: Session, message: String): Unit = {
+    try {
+      val json = mapper.readTree(message)
+      val rpcRequest = new RpcRequest(json)
+
+      rpcRequest.method match {
+        case SUBSCRIBE_REQUEST.request =>
+          subscribe(session, rpcRequest)
+        case UNSUBSCRIBE_REQUEST.request =>
+          unsubscribe(session, rpcRequest)
+        case unknownMethod =>
+          WebSocketAccountServerEndpoint.send(new RpcResponseError(rpcRequest.id,
+            new RpcError(RpcCode.MethodNotFound, s"Method $unknownMethod not supported.", "")),
+            session)
+          log.debug(s"Method $unknownMethod not supported.")
+      }
+    } catch {
+      case ex: Throwable =>
+        val sw = new StringWriter
+        ex.printStackTrace(new PrintWriter(sw))
+        WebSocketAccountServerEndpoint.send(new RpcResponseError(new RpcId("1"),
+          new RpcError(RpcCode.ExecutionError, "Websocket On receive message processing exception occurred", sw.toString)),
+          session)
+        log.debug("Websocket On receive message processing exception occurred = " + ex.getMessage)
+    }
+  }
+
+  private def subscribe(session: Session, rpcRequest: RpcRequest): Unit = {
+    val rpcParams = rpcRequest.params
+    val subscribeMethod = rpcParams.get(0)
+    val subscriptionId = createSubscriptionId()
+
+    subscribeMethod.asText() match {
+
+      case NEW_HEADS_SUBSCRIPTION.method =>
+        WebSocketAccountServerEndpoint.addNewHeadsSubscription(Subscription(session, subscriptionId))
+        log.debug("New Subscription on newHeads "+session.getId)
+        WebSocketAccountServerEndpoint.send(new RpcResponseSuccess(rpcRequest.id, subscriptionId), session)
+
+      case NEW_PENDING_TRANSACTIONS_SUBSCRIPTION.method =>
+        WebSocketAccountServerEndpoint.addNewPendingTransactionsSubscription(Subscription(session, subscriptionId))
+        log.debug("New Subscription on newPendingTransactions "+session.getId)
+        WebSocketAccountServerEndpoint.send(new RpcResponseSuccess(rpcRequest.id, subscriptionId), session)
+
+      case LOGS_SUBSCRIPTION.method =>
+        if (rpcParams.size() < 2)
+          WebSocketAccountServerEndpoint.send(new RpcResponseError(rpcRequest.id,
+            new RpcError(RpcCode.InvalidParams, "Missing filters (address, topcis).", "")),
+            session)
+        val logs = rpcParams.get(1)
+        val addressFilter = getLogAddresses(logs)
+        val topicFilter = getLogTopics(logs)
```

Comment:
1. why do we do it manually? Create a LogSubscriptionData class. Currently it is very difficult to understand what is the allowed data structure.
2. how do we detect errors? For example, address is not a valid `0x` prefixed 20 bytes hex. same for topics. In the Data class use `Address` and `Hash` types. It will make you magic.

Reply:
Done